File: app/src/main/java/com/shvants/runninglife/ui/feed/FeedFragment.java
package com.shvants.runninglife.ui.feed;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.shvants.runninglife.R;
import com.shvants.runninglife.ui.base.BaseFragment;
import com.shvants.runninglife.ui.main.MainActivity;
import com.shvants.runninglife.ui.model.UiMoveModel;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.Arrays;
import java.util.List;

import static androidx.recyclerview.widget.RecyclerView.VERTICAL;
import static com.shvants.runninglife.utils.Const.FeedFragment.TITLE;
import static java.lang.Boolean.FALSE;

public class FeedFragment extends BaseFragment {

    private RecyclerView recyclerView;
    private RecyclerView.LayoutManager layoutManager;
    private FeedPagerAdapter feedPagerAdapter;

    @Override
    public void onSaveInstanceState(@NonNull final Bundle outState) {
        //todo
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ((MainActivity) getActivity()).setActionBarTitle(TITLE);

    }

    @Override
    public View onCreateView(@NonNull final LayoutInflater inflater,
                             final ViewGroup container,
                             final Bundle savedInstanceState) {

        final View feedView = inflater.inflate(getLayoutId(), container, FALSE);

        recyclerView = feedView.findViewById(android.R.id.list);
        layoutManager = new LinearLayoutManager(getActivity(), VERTICAL, FALSE);
        recyclerView.setLayoutManager(layoutManager);

        final List<UiMoveModel> moves = getMovesFromDb();
        feedPagerAdapter = new FeedPagerAdapter(getContext(), moves);
//        recyclerView.setAdapter(feedPagerAdapter);

        return feedView;
    }

    private List<UiMoveModel> getMovesFromDb() {
        //todo from database, it is'nt?
        return Arrays.asList(
                new UiMoveModel(1, 10.00, 60L, "Morning Run", 100, 500, 130, R.drawable.move0),
                new UiMoveModel(1, 15.00, 90L, "Afternoon Run", 150, 800, 140, R.drawable.move1),
                new UiMoveModel(1, 6.00, 40L, "Night Run", 10, 200, 110, R.drawable.move2)
        );
    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_feed;
    }
}

File: app/src/main/java/com/shvants/runninglife/ui/feed/FeedFragment.kt
package com.shvants.runninglife.ui.feed

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.shvants.runninglife.R
import com.shvants.runninglife.data.web.model.WebRepository
import com.shvants.runninglife.ui.main.MainActivity
import com.shvants.runninglife.utils.Const.FeedFragment.TITLE
import com.shvants.runninglife.utils.Const.ZERO
import java.lang.Boolean.FALSE

class FeedFragment private constructor() : Fragment(), FeedContract.Fragment {

    private lateinit var presenter: FeedContract.Presenter
    private lateinit var adapter: FeedAdapter
    private lateinit var recyclerView: RecyclerView
    private lateinit var layoutManager: LinearLayoutManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        (activity as MainActivity).setActionBarTitle(TITLE)

        presenter = FeedPresenter(context, this, WebRepository.instance)
        presenter.onCreate()

        layoutManager = LinearLayoutManager(activity)
        adapter = FeedAdapter(context, presenter)
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        val feedView = inflater.inflate(getLayoutResId(), container, FALSE)

        recyclerView = feedView.findViewById(R.id.recyclerView)
        recyclerView.layoutManager = layoutManager
        recyclerView.adapter = adapter

        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {

            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)

                val totalItemCount = layoutManager.itemCount
                val visibleItemCount = layoutManager.childCount
                val firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition()

                if (totalItemCount >= MAX_VISIBLE_ITEMS) {
                    adapter.setShowLastItemAsLoading(FALSE)

                    return
                }

                if (!adapter.isLoading
                        && visibleItemCount + firstVisibleItemPosition >= totalItemCount
                        && firstVisibleItemPosition >= ZERO
                        && totalItemCount >= PAGE_SIZE) {

                    var end = totalItemCount + PAGE_SIZE

                    if (end > adapter.itemCount) {
                        end = adapter.itemCount
                    }

                    adapter.loadMoreItems(totalItemCount, end)
                }
            }
        })

        recyclerView.itemAnimator = object : DefaultItemAnimator() {
        }

        recyclerView.addItemDecoration(DividerItemDecoration(context!!,
                DividerItemDecoration.VERTICAL))

        presenter.attachView(this)

        return feedView
    }

    override fun getLayoutResId() = R.layout.fragment_feed

    companion object {
        val instance = FeedFragment()
        const val PAGE_SIZE = 2
        const val MAX_VISIBLE_ITEMS = 4
    }
}